enhancements/mco-kargs-day1-support.md

```diff
@@ -0,0 +1,347 @@
+Add kernel args to install-config and process them in the MCO bootstrap. This
+should be a straight-forward translation from the install-config into
+MachineConfig resources.
+
+#### Part 2
+For the nuts and bolts, there are a couple ways we could achieve no-extra-reboot
+kernel args support.
+
+##### Method 1 - append to `etc/pivot/kernel-args`
+
+- Update the MachineConfigServer to parse kernel args in the MachineConfigs and
+  append them in the `/etc/pivot/kernel-args` file within the ignition it
+  serves, similar to how hyperthreading can be disabled in 4.2.
+- Update the MachineConfigDaemon to compare k-args set on MachineConfigs with
+  those set via rpm-ostree to determine if there are changes requiring a reboot.
+
+Pros:
+- Requires no additional packaging in RHCOS or other tricky methods to update
+  what's available on the host.
+- Allows new nodes in clusters with 4.2 bootimages to have k-args set without
+  the additional reboot.
+- Brings the MCD closer to the kubernetes model of inspecting current state and
+  reconciling desired spec with that.
+
+Cons:
+- Further special-cases kernel args as state passed to pivot. Any other config
+  that should be set before os pivot would need to be specially handled as well.
+- Determining the current state from the host rather than the current config
+  could get tricky, and is different from how other config fields are handled.
+  See https://github.com/openshift/machine-config-operator/pull/245
+
+
+##### Method 2 - MCD-as-pivot packaged in RHCOS
+
+- Installer in 4.2 ships machine-config-daemon package in RHCOS host which handles [early pivot](https://github.com/openshift/machine-config-operator/pull/859).
+- [MCS inverts MachineConfig](https://github.com/openshift/machine-config-operator/pull/868) containing non-ignition part (like osImageURL,kernelArguments) during first boot in /etc/ignition-machine-config-encapsulated.json on hosts
+- During the first boot, MCD is patched to parse the ignition-machine-config-encapsulated.json available on the host to read and set the k-args if specified.
+- machine-config-dameon package in RHCOS is updated so that it has required changes on the host during early boot.
+- MCS is updated so that MCD first boot service runs to process kargs
+- As a result, we get kargs applied on RHCOS nodes during first boot without incurring extra reboot.
+
+Pros:
+- Moves us to a more flexible model where the MCD can be the component managing
+  the host.
+- Removes any drift between how k-args are handled on day-1 or day-2 by using
+  the same code paths.
+
+Cons:
+- Requires latest MCD package with necessary fixes in RHCOS shipped with installer, any further changes needs [updating bootimage](https://github.com/openshift/os/issues/381) which is currently not in place.
+- Will only work on clusters with bootimages at 4.3 or higher, i.e. clusters
+  installed at 4.1 or 4.2 and upgraded will not have k-args applied without a
+  double reboot on new nodes spun up.
```

Comment:
PM acked on this and we can live with this for 4.3

Reply:
Yeah I think for now let's just document that there's slowness/bugs if you have an upgraded cluster from 4.[12].

Reply:
just to understand (maybe i'm reading this wrong): are we saying that 4.1/4.2 that upgrade to 4.3 won't have this avail?
and won't be able to pick this up in an upgrade ever and will have to migrate to a fresh 4.3 install? or are we saying that those versions _will_ eventually be able to upgrade to 4.x and pick this up?

Reply:
According to my understanding, Yes, 4.1/4.2 based cluster won't have kargs day1 support available after upgrade to 4.3 because they will still be using older bootimage. We will need latest machine-config-daemon available on RHCOS host when a node boots for the first time.